### DISCUSSION TITLE
Questions about memory usage and buffers

### DISCUSSION THREAD

**Discussion by bunny-therapist** [nginx/unit#1142] [Support]
I am currently experimenting with transitioning from deployments based on Nginx (regular proxy Nginx) + separate application servers ( to just using Unit. The reason we are using Nginx is that we (due to organization requirements) need to expose management endpoints on one port and application endpoints on another port, which no python application server software could support. Unit allows us to have the port separation and it can also run our applications, which is great.

I wanted to ask some question for clarification:

1. While experimenting and comparing, I noticed that the (kubernetes) pods running Unit used more memory than the ones that use Nginx + Uvicorn. (Almost twice as much.) Is this expected? I am not great at interpreting `top` output but my impression is that the unit router process reserves a lot of memory; please correct me if I am wrong.
2. With Nginx, we regularly have logged warnings when it buffers requests to file; we usually increase the size of the buffers to silence these warnings. I have noticed that Unit produces no such warnings and that there are also no buffer-size configuration options. How does this actually work? Are requests not buffered? Is there a fixed-size buffer that is huge? I would love it to just not have to care about these buffer warnings, but I would like to understand why we have this problem in Nginx but not with Unit.

Thank you very much. I am very interested in Unit and happy I discovered it.

**Comment by hongzhidao** [comment]
Hi,
About the second question. It's related to how to process the request body.
In nginx, there is an option to enable request body buffering or not.
https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_request_buffering
It's useful for server-side processing status.
But in Unit, we thought it's better to process the request body as soon as possible, so we save it into a temporary file if it's too big.

> Are requests not buffered?

The request body is buffered.

> Is there a fixed-size buffer that is huge

If it's too large, it's dumped into a temporary file.

> How does this actually work? 

There is a buffer of 16K size to save the request body, if the body size is larger than that, it will be saved into a temporary file.
Note the max body size is 8M.

And you can set their size in the configuration.
https://unit.nginx.org/configuration/#settings
